components/tooltip/script.js

```diff
@@ -0,0 +1,97 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Coherent Labs AD. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import components from 'coherent-gameface-components';
+import template from './template.html';
+const TOOLTIP_MARGIN = 10;
+
+class Tooltip extends HTMLElement {
+    constructor() {
+        super();
+        this.template = template;
+
+        this.state = {
+            visible: false
+        };
+    }
+
+    connectedCallback() {
+        this.position = this.getAttribute('position') || 'top';
+        this.showOn = this.getAttribute('on') || 'click';
+        this.hideOn = this.getAttribute('off') || 'click';
+        this.elementSelector = this.getAttribute('for');
+        // TODO: check if this.elementSelector is valid and attached to DOM
+        this.triggerElement = document.querySelector(this.elementSelector);
+
+        components.loadResource(this)
+            .then((result) => {
+                this.template = result.template;
+                components.renderOnce(this);
+                this.attachEventListeners();
+            })
+            .catch(err => console.error(err));
+    }
+
+    attachEventListeners() {
+        // TODO: validate if showOn is a valid onEvent handler
+        if (this.showOn === this.hideOn) {
+            this.triggerElement.addEventListener(this.showOn, (e) => this.toggle(e));
+            return;
+        }
+
+        this.triggerElement.addEventListener(this.showOn, (e) => this.show(e));
+        this.triggerElement.addEventListener(this.hideOn, (e) => this.hide(e));
+    }
+
+    toggle() {
+        if (this.state.visible) {
+            this.hide();
+        } else {
+            this.show();
+        }
+
+        this.state.visible = !this.state.visible;
+    }
+
+    hide () {
+        this.style.display = 'none';
+    }
+
+    show() {
+        // use visibility before showing to calculate the size
+        this.style.visibility = 'hidden';
+        this.style.display = '';

```

Comment:
We have a `.hidden-tooltip` class in the style.css but we don't use it. Can we add one for switching the display and switch the visibility with `classList` instead of using `style`.

Reply:
My original idea was to use class list and I did it with class list, but because we add the class on the <gameface-tooltip> in the connected callback or in the constructor - there is a brief moment where the tooltip is visible and then it is hidden and it blinks. This is why I switched to inline style and I forgot to remove the class.
If you know how to fix the blinking without forcing the customer to add the class, please share. I couldn't think of other solution.

Reply:
Does adding the hidden class to the wrapper element inside the template by default is fixing the blinking problem? 
```
//template.html
<div class="wrapper hidden">
...
</div>
```
Then just work with the wrapper element class list to remove or add the hidden class.

Reply:
No, because it takes some time for the template to be loaded and rendered during which the content is visible:

    <gameface-tooltip id="tutorial" target=".manual" on="click" position="top" off="click">
    everything here is visible until render
    </gameface-tooltip>